### ISSUE TITLE
DTensor projection reshapes prevent batched FFN and attention fusion

### ISSUE THREAD

**Issue by Boffee** [Boffee/piper-kernels#109]
The flatten/restore operations introduced around DTensor linears prevent both **batched SwiGLU FFN fusion** and **sparse attention projection fusion**. The affected paths retain intermediates that the existing fused kernels would avoid. This affects ConvRot INT8, Piper NVFP4, and ConvRot NVFP4.

The intended H3 integration runs fused kernels on local shards: attention is sharded by heads, FFNs by intermediate width, and partial outputs are summed after the corresponding output/down projection. DTensor boundaries around these local computations must preserve fusion. A native DTensor sharding strategy inside the attention operator is **not required for this design and is not part of this issue's acceptance criteria**.

Reproduced after the linear fix for #106, at commit `a133701cd9af43595ed782cdd6baa7238837fee0`. The original transpose/linear failures are resolved in that checkout; this is a subsequent fusion-recognition gap.

### FFN behavior

| Format | Local `[1, 256, 256]` | DTensor `[256, 256]` | DTensor `[1, 256, 256]` |
| --- | --- | --- | --- |
| ConvRot INT8 | Fused | Fused | Three separate linears |
| Piper NVFP4 | Fused | Fused | Three separate linears |
| ConvRot NVFP4 | Fused | Fused | Three separate linears |

The DTensor comparison reproduces with one ordinary Gloo rank, replicated CUDA weights/activations, and fullgraph Inductor compilation with the appropriate Piper FFN compile options. It does not require Piper Offload or a custom process group.

Separate two-rank tests with gate/value output sharding and down-projection input sharding show the same difference, including with streamed weights. Flattened cases exactly match independent local **fused** references. Batched cases exactly match local compiled **unfused** references. This is a missed optimization, not an observed incorrect unfused result. It retains full FFN intermediates instead of selecting the existing chunked-workspace implementation, which matters for diffusion activation-memory usage.

### FFN graph difference

For the batched DTensor case, each projection is surrounded by flatten/restore operations. The captured INT8 graph is structurally:

```text
x: [1, 256, 256]
gate = reshape(linear(reshape(x, [256, 256]), gate_weight), [1, 256, 512])
value = reshape(linear(reshape(x, [256, 256]), value_weight), [1, 256, 512])
hidden = silu(gate) * value
output = reshape(linear(reshape(hidden, [256, 512]), down_weight), [1, 256, 256])
```

The fusion matcher currently expects the gate/value/down linears to connect directly through the SwiGLU expression. These reshapes prevent that pattern from matching. The existing fused GPU implementations already flatten leading dimensions internally and restore the output shape.

### Attention behavior

Additional compilation probes at the same commit reproduce the following for **all three formats**:

| Projection setup | QKV preparation fusion | Attention/output-projection fusion |
| --- | --- | --- |
| Local quantized tensors | Selected | Selected |
| DTensor QKV linears | Missed | Missed downstream |
| DTensor output linear only | Selected | Missed |
| DTensor QKV and output linears | Missed | Missed |

The tested attention region uses `[1,192,256]` activations, two 128-wide heads, per-head Q/K RMSNorm, 96-wide partial RoPE, mean-routing sparse Piper attention, and a 256-wide output projection. NVFP4 QKV activation quantization is dynamic; NVFP4 output activation quantization is **static**, so the local baseline satisfies the existing output-fusion requirements.

These are isolated ordinary singleton DTensor tests using a standard Gloo/CUDA mesh. Projection weights and inputs are replicated DTensors; projection outputs are explicitly unwrapped before local attention. The output-projection input is wrapped again when testing that boundary. This tests the local-attention integration without requiring a DTensor strategy for the core attention custom op. It is not a completed multi-rank head-sharding test.

The local INT8 graph selects `convrot_int8_sparse_piper_project_key`, `convrot_int8_sparse_piper_project_value`, and `convrot_int8_sparse_piper_projected_query_attention_output`. With DTensor QKV linears, the graph instead contains separate linears, materialized norm/RoPE, and the core `sparse_piper_attention` operator. With only the output linear using DTensor, QKV preparation remains fused, but `sparse_piper_attention_from_quantized` feeds a separate output linear.

The problematic graph includes these additional wrappers:

```text
Q/K: input [B,T,C] -> flatten [B*T,C] -> linear
     -> restore [B,T,H*D] -> head reshape [B,T,H,D] -> norm/RoPE
V:   input [B,T,C] -> flatten [B*T,C] -> linear
     -> restore [B,T,H*D] -> head reshape [B,T,H,D]
Out: attention [B,T,H,D] -> flatten heads [B,T,H*D]
     -> flatten rows [B*T,H*D] -> linear -> restore [B,T,C]
```

To reproduce using the canonical attention regions in the existing fusion tests, wrap the selected projection weights with `DTensor.from_local(..., [Replicate()])`, and use these projection boundaries inside a fullgraph-compiled forward:

```python
# QKV boundary; norms, RoPE, and core attention consume ordinary local tensors.
dx = DTensor.from_local(hidden_states, mesh, [Replicate()], run_check=False)
q, k, v = query(dx).to_local(), key(dx).to_local(), value(dx).to_local()
# Apply the existing per-head reshape, RMSNorm/RoPE and sparse-attention code.

# Output boundary, after local attention and the normal head-flattening reshape.
dy = DTensor.from_local(attention_output.flatten(2), mesh, [Replicate()], run_check=False)
result = output_projection(dy).to_local()
```

Use the corresponding `*_sparse_piper_compile_options` and capture the post-pass operator targets, as in the FFN reproduction below. Test QKV-only, output-only, and both boundaries against the local baseline. Relevant fixtures are in `tests/fusions/{convrot_int8,nvfp4,convrot_nvfp4}_sparse_piper/test_compile.py`.

The isolated attention runs returned finite outputs of the expected shape; this establishes operator selection, not numerical equivalence. Manual local flatten/restore controls also miss output fusion for all formats. The simpler manual QKV control is already accepted by the NVFP4 matchers, while the actual DTensor graph is not, so support must recognize equivalent projection graphs rather than indiscriminately strip reshapes.

### Standalone FFN reproduction

This compact INT8 example covers both local and ordinary DTensor execution. The broader validation above also covered both NVFP4 formats.

Environment: RTX 5090, Linux, Python 3.14, PyTorch `2.13.0+cu130`, TorchAO `0.18.0+cu130`.

Run from the Kernels checkout with dependencies installed:

```sh
PYTHONPATH=src TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 python repro.py
```

Compilation caches were disabled to isolate the source change from previously compiled artifacts.

```python
import uuid
import torch
import torch.distributed as dist
from torch import nn
from torch.nn import functional as F
from torch._inductor.custom_graph_pass import CustomInferenceAwareGraphPass
from torch.distributed.device_mesh import DeviceMesh
from torch.distributed.tensor import DTensor, Replicate
from piper_kernels.linear.convrot.int8 import ConvRotInt8Tensor
from piper_kernels.fusions.convrot_int8_swiglu_ffn import convrot_int8_swiglu_ffn_compile_options


class FFN(nn.Module):
    def __init__(self):
        super().__init__()
        for name, out_features, in_features in (
            ("gate", 512, 256), ("value", 512, 256), ("down", 256, 512),
        ):
            source = torch.randn(out_features, in_features, device="cuda", dtype=torch.bfloat16) * 0.02
            layer = nn.Linear(in_features, out_features, bias=False, device="meta")
            layer.weight = nn.Parameter(ConvRotInt8Tensor.from_hp(source, group_size=64), False)
            setattr(self, name, layer)

    def forward(self, x):
        return self.down(F.silu(self.gate(x)) * self.value(x))


class Capture(CustomInferenceAwareGraphPass):
    def __init__(self):
        self.targets = []
        self.identity = uuid.uuid4().bytes

    def __call__(self, graph, is_inference):
        assert is_inference
        self.targets = [str(n.target) for n in graph.nodes if n.op == "call_function"]

    def uuid(self):
        return self.identity


torch.set_num_threads(1)
torch.manual_seed(106)
dist.init_process_group("gloo", store=dist.HashStore(), rank=0, world_size=1)
try:
    mesh = DeviceMesh("cuda", [0])
    with torch.no_grad():
        for use_dtensor in (False, True):
            for shape in ((256, 256), (1, 256, 256)):
                torch.compiler.reset()
                model = FFN()
                x = torch.randn(shape, device="cuda", dtype=torch.bfloat16)
                if use_dtensor:
                    for layer in model.children():
                        layer.weight = nn.Parameter(DTensor.from_local(
                            layer.weight.detach(), mesh, [Replicate()], run_check=False), False)
                    x = DTensor.from_local(x, mesh, [Replicate()], run_check=False)
                capture = Capture()
                options = convrot_int8_swiglu_ffn_compile_options(
                    {"triton.cudagraphs": False, "compile_threads": 1})
                passes = options["post_grad_custom_pre_pass"]
                options["post_grad_custom_pre_pass"] = (*passes, capture)
                torch.compile(model, fullgraph=True, options=options)(x)
                print("DTensor:", use_dtensor, "shape:", shape,
                      "fused FFNs:", capture.targets.count("piper_kernels.convrot_int8_swiglu_ffn.default"),
                      "separate linears:", capture.targets.count("piper_kernels.convrot_int8_linear.default"),
                      flush=True)
finally:
    dist.destroy_process_group()
```

Observed output:

```text
DTensor: False shape: (256, 256) fused FFNs: 1 separate linears: 0
DTensor: False shape: (1, 256, 256) fused FFNs: 1 separate linears: 0
DTensor: True shape: (256, 256) fused FFNs: 1 separate linears: 0
DTensor: True shape: (1, 256, 256) fused FFNs: 0 separate linears: 3
```

### Expected behavior and suggested scope

- Preserve the existing chunked SwiGLU FFN fusion for batched DTensor projections in all three formats.
- Preserve attention QKV preparation and attention/output-projection fusion across the corresponding DTensor/local-tensor boundaries in all three formats, for configurations supported by each existing fused kernel.
- Prefer shared handling of equivalent `reshape -> Piper linear -> reshape` projection patterns, usable by both FFN and attention matchers. Existing GPU kernels can be reused; shared normalization is a suggested implementation approach.
- Normalize only shapes that preserve element order and the feature axis. Preserve dtypes, quantization metadata and externally consumed intermediates, and retain appropriate layout/consumer guards.
- Add 2-D and batched regression coverage, including symbolic leading dimensions. For attention, test QKV and output boundaries separately and together. Assert the fused operator is actually selected, since output correctness alone can pass while fusion is absent.
- Validate the intended local-shard boundaries: attention heads and FFN intermediate features are partitioned, fused computation stays local, and SUM combines partial outputs after the output/down projection. Keep collectives outside the local fused region.
- Compare fused results with the corresponding local fused quantized reference. Fused and unfused implementations can have different quantization boundaries and need not be bitwise equal.

Callers should keep standard DTensor projection plans and local shard kernels without manually flattening model activations solely to trigger fusion. This issue does not require implementing native DTensor dispatch inside the core attention custom operator.
